File: greedy/knapsack_problem/python/Knapsack.py
# Problem Statement: Given a Knapsack of a maximum capacity of W and N items each with its own value and weight, 
# throw in items inside the Knapsack such that the final contents has the maximum value.
# 
# Credits to http://www.geeksforgeeks.org/knapsack-problem/ for the best-looking solution.

# Main method
def main():
    # Stores the value of objects
    value = [10, 40, 30, 50]

    # Stores the weight of objects
    weight = [5, 4, 6, 3]

    # Stores the capacity of the knapsack
    W = 10

    # gets total number of objects
    n = len(value)

    print(knapSack(W, weight, value, n))


# A Dynamic Programming based Python Program for 0-1 Knapsack problem
# Returns the maximum value that can be put in a knapsack of capacity W,
# wt array and val array, and n is the total number of objects
def knapSack(W, wt, val, n):
    
    # creates 2D array
    K = [[0 for x in range(W + 1)] for y in range(n + 1)]
 
    # Build table K[][] in bottom up manner
    for i in range(n + 1):
        for w in range(W + 1):
            if i == 0 or w == 0:
                K[i][w] = 0
            elif wt[i - 1] <= w:
                K[i][w] = max(val[i-1] + K[i-1][w-wt[i-1]],  K[i-1][w])
            else:
                K[i][w] = K[i-1][w]
 
    return K[n][W]

main()
 

File: greedy/knapsack_problem/java/Knapsack.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Stack;

/**
 * Problem Statement: Given a Knapsack of a maximum capacity of W and N items each with its own value and weight, 
 * throw in items inside the Knapsack such that the final contents has the maximum value.
 *
 * Credits to http://rerun.me/2014/05/27/the-knapsack-problem/ for the clean solution.
 */

public class Knapsack {

    public static void main(String[] args) throws Exception {
        int value[] = {10, 40, 30, 50};
        int weight[] = {5, 4, 6, 3};
        int W = 10;

        System.out.println(knapsack(value, weight, W));
    }

    public static int knapsack(int value[], int weights[], int W) {
        
        int N = weights.length;
        
        int[][] dp = new int[N + 1][W + 1];

        for (int item = 1; item <= N; item++) {
            for (int weight = 1; weight <= W; weight++) {

                //Is the current item's weight less
                //than or equal to running weight
                if (weights[item - 1] <= weight) {

                    //Given a weight, check if the value of the current item + value of the item that we could afford
                    // with the remaining weight is greater than the value without the current item itself
                    dp[item][weight] = Math.max(
                            value[item - 1] + dp[item - 1][weight - weights[item - 1]],
                            dp[item - 1][weight]);
                } else {
                    //If the current item's weight is more than the
                    //running weight, just carry forward the value
                    //without the current item
                    dp[item][weight] = dp[item - 1][weight];
                }
            }

        }

        //Printing the matrix of values which we calculated.
        for (int[] rows : dp) {
            for (int col : rows) {
                System.out.format("%5d", col);
            }
            System.out.println();
        }

        return dp[N][W];
    }

}
